### DISCUSSION TITLE
Finalize interrupted downloaded HLS video stream

### DISCUSSION THREAD

**Discussion by V33m** [aclap-dev/video-downloadhelper#1295] [Help and Q&A]
### Your question

Sometimes, Firefox or the VDH add-on (or its companion app) may crash while downloading a file. My question is: how can I finalize a partially downloaded file and make it playable? I am using the latest versions of both Firefox and VDH.

Here’s the situation:

I was downloading a file named `test.mp4`, which uses HLS, MP4, AAC, and H264 formats. During the process, the download was interrupted. Now, I’m left with an incomplete file that isn’t playable in VLC, and when I check it in Mediainfo, it shows:

```
Format                                   : MPEG-4
Format profile                           : Base Media
Codec ID                                 : isom (isom/iso2/avc1/mp41)
File size                                : 8.07 GiB
```

I have ffmpeg installed and am looking for a command or method to finalize the file. Since it's a stream, redownloading isn't an option. The file is either recoverable, or it's lost. Any suggestions on how to complete it?

### Addon & CoApp version and details

version: 9.1.0.44
target: mozilla
channel: stable
lang: en-US
coapp: {"found":true,"path":"/Applications/net.downloadhelper.coapp.app/Contents/MacOS/vdhcoapp","version":"2.0.19","new_version":false}
platform: aarch64 mac
UA: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:130.0) Gecko/20100101 Firefox/130.0

**Comment by GrampaWildWilly** [comment]
I do not work for VDH.  I'm just another user.

The file is almost certainly lost.  With MP4 videos, there is a block of data called the moov atom.  That block of data is not written to the file until the download completes.  Without that data, the file is useless.  I tried a few repair tools on such a file a while ago & they all failed, claiming they couldn't find the moov atom.  Perhaps someone with more knowledge than I have can tell you better news.  But unfortunately, I have only bad news for you.  I would have told you to just download the file again, but you said that's not possible.  So I think you're stuck.

If your target file is MKV, this is not a problem.  It is my understanding that the structure of MKV is such that not only is it continuously a "complete" file, but you can actually play what you have while more of it is being recorded.  You might consider preferring MKV if that is an option.  I'm not sure it always is, but that might be something to keep in mind.

**Comment by V33m** [comment]
Thank you so much for reaching out! Then I'll consider the file lost. Speaking of MKV and VDH, I guess the option 'Download and convert to MKV' still won't resolve the case here as it downloads the file as MP4 and then converts to MKV after the download is complete. Let me know if there is an option to try to download the file and save it as MKV directly

**Reply by GrampaWildWilly** [reply]
The convert functions of VDH that were in older releases have not been ported forward to VDH 9.  Converting would have to be done outside VDH.  As for MKV, VDH lists MKV variants . . . sometimes.  That mostly depends on the web site & what it is offering.  YouTube usually offers MKV variants.  But elsewhere, every site makes its own decisions.  You are correct, that converting after the fact would be no help.  You just have to hope your connection doesn't drop in the middle of a livestream.  I've had that happen.  There's not much you can do about it.  That sort of thing is pretty much out of our control.
